core/drivers/sam/at91_audio_pll.c

```diff
@@ -0,0 +1,420 @@
+
+	DMSG("A PLL: fr = %llu", fr);
+
+	fr = UDIV_ROUND_NEAREST(fr, AUDIO_PLL_DIV_FRAC);
+
+	DMSG("A PLL: fr = %llu", fr);
+
+	return parent_rate * (nd + 1) + fr;
+}
+
+static unsigned long clk_audio_pll_frac_get_rate(struct clk *clk,
+						 unsigned long parent_rate)
+{
+	struct clk_audio_frac *frac = clk->priv;
+	unsigned long fout = 0;
+
+	fout = clk_audio_pll_fout(parent_rate, frac->nd, frac->fracr);
+
+	DMSG("A PLL: %s, fout = %lu (nd = %u, fracr = %lu)\n", __func__,
+	     fout, frac->nd, (unsigned long)frac->fracr);
+
+	return fout;
+}
+
+static unsigned long clk_audio_pll_pad_get_rate(struct clk *clk,
+						unsigned long parent_rate)
+{
+	struct clk_audio_pad *apad_ck = clk->priv;
+	unsigned long apad_rate = 0;
+
+	if (apad_ck->qdaudio && apad_ck->div)
+		apad_rate = parent_rate / (apad_ck->qdaudio * apad_ck->div);
+
+	DMSG("A PLL/PAD: %s, apad_rate = %lu (div = %u, qdaudio = %u)\n",
+	     __func__, apad_rate, apad_ck->div, apad_ck->qdaudio);
+
+	return apad_rate;
+}
+
+static unsigned long clk_audio_pll_pmc_get_rate(struct clk *clk,
+						unsigned long parent_rate)
+{
+	struct clk_audio_pmc *apmc_ck = clk->priv;
+	unsigned long apmc_rate = 0;
+
+	apmc_rate = parent_rate / (apmc_ck->qdpmc + 1);
+
+	DMSG("A PLL/PMC: %s, apmc_rate = %lu (qdpmc = %u)\n", __func__,
+	     apmc_rate, apmc_ck->qdpmc);
+
+	return apmc_rate;
+}
+
+static TEE_Result clk_audio_pll_frac_compute_frac(unsigned long rate,
+						  unsigned long parent_rate,
+						  unsigned long *nd,
+						  unsigned long *fracr)
+{
+	unsigned long long tmp = 0;
+	unsigned long long rem = 0;
+
+	if (!rate)
+		return -1;
+
+	tmp = rate;
+	rem = tmp % parent_rate;
+	tmp /= parent_rate;
+	if (!tmp || tmp >= AUDIO_PLL_ND_MAX)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	*nd = tmp - 1;
+
+	tmp = rem * AUDIO_PLL_DIV_FRAC;
+	tmp = UDIV_ROUND_NEAREST(tmp, parent_rate);
+	if (tmp > AT91_PMC_AUDIO_PLL_FRACR_MASK)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	/* we can cast here as we verified the bounds just above */
+	*fracr = (unsigned long)tmp;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result clk_audio_pll_frac_set_rate(struct clk *clk,
+					      unsigned long rate,
+					      unsigned long parent_rate)
+{
+	struct clk_audio_frac *frac = clk->priv;
+	unsigned long fracr = 0;
+	unsigned long nd = 0;
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	DMSG("A PLL: %s, rate = %lu (parent_rate = %lu)\n", __func__, rate,
+	     parent_rate);
+
+	if (rate < AUDIO_PLL_FOUT_MIN || rate > AUDIO_PLL_FOUT_MAX)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	res = clk_audio_pll_frac_compute_frac(rate, parent_rate, &nd, &fracr);
+	if (res)
+		return res;
+
+	frac->nd = nd;
+	frac->fracr = fracr;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result clk_audio_pll_pad_set_rate(struct clk *clk,
+					     unsigned long rate,
+					     unsigned long parent_rate)
+{
+	struct clk_audio_pad *apad_ck = clk->priv;
+	uint8_t tmp_div = 1;
+
+	DMSG("A PLL/PAD: %s, rate = %lu (parent_rate = %lu)\n", __func__, rate,
```

Comment:
drop argument `__func__` adn trailing `\n`, already handled by `DMSG()`.

Reply:
Done for all DMSG